src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20240103_143700.java

```diff
@@ -0,0 +1,497 @@
+            log.error("Migration will be skipped and marked run because the local VC base URL is not configured.");
+            return;
+        }
+
+        if (bitbucketLocalVCMigrationService.get().getDefaultBranch() == null) {
+            log.error("Migration will be skipped and marked run because the default branch is not configured.");
+            return;
+        }
+
+        var programmingExerciseCount = programmingExerciseRepository.count();
+        var studentCount = programmingExerciseStudentParticipationRepository.findAllWithRepositoryUri(Pageable.unpaged()).getTotalElements();
+
+        if (programmingExerciseCount == 0) {
+            // no exercises to change, migration complete
+            return;
+        }
+
+        log.info("Will migrate {} programming exercises and {} student repositories now. This might take a while", programmingExerciseCount, studentCount);
+
+        final long totalFullBatchCount = programmingExerciseCount / BATCH_SIZE;
+        final long threadCount = Math.max(1, Math.min(totalFullBatchCount, MAX_THREAD_COUNT));
+        final long estimatedTimeExercise = getRestDurationInSeconds(0, programmingExerciseCount, 3, threadCount);
+        final long estimatedTimeStudents = getRestDurationInSeconds(0, studentCount, 1, threadCount);
+
+        final long estimatedTime = (estimatedTimeExercise + estimatedTimeStudents);
+        log.info("Using {} threads for migration, and assuming {}s per repository, the migration should take around {}", threadCount, ESTIMATED_TIME_PER_REPOSITORY,
+                TimeLogUtil.formatDuration(estimatedTime));
+
+        // Use fixed thread pool to prevent loading too many exercises into memory at once
+        ExecutorService executorService = Executors.newFixedThreadPool((int) threadCount);
+
+        /*
+         * migrate the solution participations first, then the template participations, then the student participations
+         */
+        AtomicInteger solutionCounter = new AtomicInteger(0);
+        final var totalNumberOfSolutions = solutionProgrammingExerciseParticipationRepository.count();
+        log.info("Found {} solution participations to migrate.", totalNumberOfSolutions);
+        for (int currentPageStart = 0; currentPageStart < totalNumberOfSolutions; currentPageStart += BATCH_SIZE) {
+            Pageable pageable = PageRequest.of(currentPageStart / BATCH_SIZE, BATCH_SIZE);
+            var solutionParticipationPage = solutionProgrammingExerciseParticipationRepository.findAll(pageable);
+            log.info("Will migrate {} solution participations in batch.", solutionParticipationPage.getNumberOfElements());
+            executorService.submit(() -> {
+                migrateSolutions(solutionParticipationPage.toList());
+                solutionCounter.addAndGet(solutionParticipationPage.getNumberOfElements());
+                logProgress(solutionCounter.get(), totalNumberOfSolutions, threadCount, 2, "solution");
```

Comment:
I suppose this is right, but why are there two repos per solution and 1 per  template/student participation?

Reply:
because with the solution participation we also migrate the test repo